astropy/units/quantity.py

```diff
@@ -507,6 +510,12 @@ def __new__(cls, value, unit=None, dtype=np.inexact, copy=True, order=None,
         value = np.array(value, dtype=dtype, copy=copy, order=order,
                          subok=True, ndmin=ndmin)
 
+        # For no-user-input unit, make sure the constructed unit matches the
+        # structure of the data.
+        if UNIT_IS_NONE and is_structured(value):
```

Comment:
This does not remove the possibility of having a structured value with an unstructured unit. But at least it will now be the user's fault.